Set up API test project with Cucumber and Axios

AnnaBileva24 opened the pull request:
- Configured Cucumber for behavior-driven development (BDD).
- Integrated Axios for making API calls.
- Added feature files.
- Implemented step definitions for handling test steps.
- Created utility function in `utils/apiUtils.js`, including `checkUserExists` for user existence validation.

AnnaBileva24 committed:
Set up API test project with Cucumber and Axios

- Configured Cucumber for behavior-driven development (BDD).
- Integrated Axios for making API calls.
- Added feature files.
- Implemented step definitions for handling test steps.
- Created utility function in `utils/apiUtils.js`, including `checkUserExists` for user existence validation.

fedor-bilev commented on features/create_user.feature:
```diff
@@ -0,0 +1,12 @@
+        Then the user is successfully created
+        And the response contains the following user details
+        | name    | John      |
+        | job     | Developer |
```
Consider adding some negative test cases, e.g.
```
Scenario: Attempt to create a user with invalid data
  When I send a POST request to "/users" with invalid data
  Then the response status should be 400
  And the response body should contain an error message
```

fedor-bilev commented on features/step_definitions/stepdefs.js:
```diff
@@ -0,0 +1,72 @@
+});
+
+When("a list of users on page {int} is requested", async function (page) {
+  const resp = await axios.get(`https://reqres.in/api/users?page=${page}`);
```
Would be nice to extract all call to the server to a separate util client e.g.: `ReqresClient.js` and store all the rest logic there

```js
const axios = require('axios');

const apiClient = axios.create({
    baseURL: 'https://reqres.in/api',
    timeout: 1000,
});

const createUser = async (userData) => {
    try {
        const response = await apiClient.post('/users', userData);
        console.log(`User created with ID: ${response.data.id}`);
        return response.data;
    } catch (error) {
        console.error('Error creating user:', error);
        throw error;
    }
};

const getUser = async (userId) => {
    try {
        const response = await apiClient.get(`/users/${userId}`);
        console.log(`User fetched with ID: ${userId}`);
        return response.data;
    } catch (error) {
        console.error(`Error getting user with ID ${userId}:`, error);
        throw error;
    }
};

const updateUser = async (userId, userData) => {
    try {
        const response = await apiClient.put(`/users/${userId}`, userData);
        console.log(`User updated with ID: ${userId}`);
        return response.data;
    } catch (error) {
        console.error(`Error updating user with ID ${userId}:`, error);
        throw error;
    }
};

const deleteUser = async (userId) => {
    try {
        const response = await apiClient.delete(`/users/${userId}`);
        console.log(`User deleted with ID: ${userId}`);
        return response.status === 204;
    } catch (error) {
        console.error(`Error deleting user with ID ${userId}:`, error);
        throw error;
    }
};

module.exports = {
    createUser,
    getUser,
    updateUser,
    deleteUser,
};
```

then you can use it in your step definitions
```js
const { createUser, getUser, updateUser, deleteUser } = require('./apiClient');
```

Please consider reading [SLAP](https://medium.com/@outfunnel/the-single-level-of-abstraction-principle-52b5d56ef54b)
- TLDR; To ensure clarity and maintain a consistent level of abstraction, each method should encapsulate a single level of detail. This mirrors real-life conversations where we maintain a consistent detail level. For instance, when discussing my weekend, I would mention highlights like visiting a city or enjoying a nice dinner, but not minute actions like locking the door when I left my apartment. This approach keeps the code clean and focused.

AnnaBileva24 committed:
Fix according to the comment:

1. Created utility functions in `utils/apiUtils.js to ensure clarity and maintain a consistent level of abstraction.
2. Used these functions in implemented step_definitions

fedor-bilev commented on features/step_definitions/stepdefs.js:
```diff
@@ -1,20 +1,26 @@
 
 When("a list of users on page {int} is requested", async function (page) {
-  const resp = await axios.get(`https://reqres.in/api/users?page=${page}`);
+  const resp = await await getUsersList(page);
```
Not sure if we need await here, since getUsersList already has one.
`await await getUsersList(page);` -> `getUsersList(page);`

Same for others

fedor-bilev replied:
Also could be inlined
```js
this.listOfUsers = getUsersList(page)
```

fedor-bilev commented on features/step_definitions/stepdefs.js:
```diff
@@ -50,7 +56,7 @@ When(
   async function (id, dataTable) {
     const data = dataTable.rowsHash();
-    const resp = await axios.put(`https://reqres.in/api/users/${id}`, {
+    const resp = await updateUser(id, {
```
Can't we just use `updateUser(id, dataTable.rowsHash())`?

fedor-bilev commented on features/utils/apiUtils.js:
```diff
@@ -1,14 +1,69 @@
+});
+
+const getUsersList = async (page) => {
+  try {
```
```js
  try {
    if (page) {
     return await apiClient.get(`/users?page=${page}`);
    } 
    return await apiClient.get(`/users`);
  } catch (error) {
    console.error(`Error getting list of users:`, error);
    throw error;
  }
```

fedor-bilev replied:
Not sure about `return await`

AnnaBileva24 replied:
What is the purpose of this change?

fedor-bilev replied:
Code style and readability. 
Which version is more readable?

AnnaBileva24 replied:
I find it more readable and convenient to assert when I return response

fedor-bilev commented on features/utils/apiUtils.js:
```diff
@@ -1,14 +1,69 @@
+
+const checkUserExists = async (userId) => {
+  try {
+    const response = await apiClient.get(`/users/${userId}`);
```
Try `return await apiClient.get(`/users/${userId}`);`

fedor-bilev commented on features/utils/apiUtils.js:
```diff
@@ -1,14 +1,69 @@
-const axios = require('axios');
```
Better to create a separate file for the reqres client. e.g. `reqres_client.js`

fedor-bilev replied:
The suggestion to create a separate file for the reqres_client.js is from the best practices in software development:

- Modularity: Keeping the code modular by separating concerns into different files makes it easier to maintain and understand. Each file can focus on a specific aspect of the application, reducing complexity.

- Reusability: By isolating the reqres client in its own file, you can reuse the same client across different parts of your application without duplicating code. This promotes DRY (Don't Repeat Yourself) principles.

- Testing: It's easier to test individual components when they are separated into different files. A dedicated file for the reqres client means you can write and run tests specifically for your API interactions without interference from other parts of your application.

- Collaboration: When working in a team, having separate files can reduce merge conflicts and make it clearer who is working on what. It's easier to manage version control when changes are isolated to specific areas.

By separating the reqres_client.js into its own file, you’re structuring your codebase in a way that is scalable, maintainable, and easier to navigate. This approach can lead to better software architecture and easier management of the project as it grows.

AnnaBileva24 committed:
Improved code structure and readability

1. Renamed util file to "reqres_client.js"
2. Fixed util functions to improve readability

AnnaBileva24 committed:
Optimized code in stepdefs.js
